ql/instruments/forwardrateagreement.cpp

```diff
@@ -88,24 +152,36 @@ namespace QuantLib {
     void ForwardRateAgreement::performCalculations() const {
         calculateAmount();
 
-        Handle<YieldTermStructure> discount =
-            discountCurve_.empty() ? index_->forwardingTermStructure() : discountCurve_;
+        if (!discountCurve_.empty() || index_) {
+
+            Handle<YieldTermStructure> discount =
+                discountCurve_.empty() ? index_->forwardingTermStructure() : discountCurve_;
 
-        NPV_ = amount_ * discount->discount(valueDate_);
+            NPV_ = amount_ * discount->discount(valueDate_);
+        } else
+            NPV_ = amount_ * referenceRate_.discountFactor(valueDateTime_);
```

Comment:
This is surely very unlikely to be right. The discount factor here describes how much a unit of currency at the FRA's value date is worth at the reference rate (today). But your reference rate applies to the term of the FRA (value date to maturity date), doesn't it? You can't just take a rate for the latter period and use it for the former period.

Reply:
In my intention here with the `referenceRate` I was meaning the index interest rate the FRA contract rate (`strikeForwardRate`) is compared against in order to determine the settlement amount. This comparison happens at the settlement date. The FRA effectively ends with the settlement date, as there is no longer any contractual engagement between the two counter parties once the settlement amount has been paid. The contract period is merely one of the calculation parameters used to determine the settlement amount. So if for instance we have a 1x3 FRA where `evaluationDate` is 31 January 2023, `valueDate` is 28 February 2023, `maturityDate` is 31 May 2023 then the Present Value at the `evaluationDate` is the FRA amount (the future cash flow) calculated at the `valueDate` multiplied by the discount factor. Anyway, as we said if the index is not provided then the `referenceRate`will be taken form the `Handle<YieldTermStructure>`.